### ISSUE TITLE
map reduce failed because of commons config version in 1.9.3-rc2

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#1065]
After running continuous ingest against 1.9.3-rc2 the verification map reduce job failed to start because of the following error.

```
Exception in thread "main" java.lang.NoSuchMethodError: org.apache.commons.configuration.MapConfiguration.<init>(Ljava/util/Properties;)V
        at org.apache.accumulo.core.conf.Property.precomputeDefaultValue(Property.java:976)
        at org.apache.accumulo.core.conf.Property.<clinit>(Property.java:1353)
        at org.apache.accumulo.core.client.ClientConfiguration$ClientProperty.<clinit>(ClientConfiguration.java:70)
        at org.apache.accumulo.core.cli.ClientOpts.updateKerberosCredentials(ClientOpts.java:255)
        at org.apache.accumulo.core.cli.ClientOpts.updateKerberosCredentials(ClientOpts.java:247)
        at org.apache.accumulo.core.cli.ClientOpts.parseArgs(ClientOpts.java:284)
        at org.apache.accumulo.test.continuous.ContinuousVerify.run(ContinuousVerify.java:172)
        at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:76)
        at org.apache.accumulo.test.continuous.ContinuousVerify.main(ContinuousVerify.java:233)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:244)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:158)
```

Investigated and found the following :
  * Hadoop 2.9.2 used commons config version 1.6
  * Accumulo switched to commons config version 1.10 after 1.9.2 in commit 06d80292d869edc495bf691c0d43be52dc0dcdca for #659
 * The new method `MapConfiguration(Properties props)` was added in commons config 1.8.
 * There was an existing method `MapConfiguration(Map<String,?> map)` that our code used to bind to.  Now [Property.java line 976](https://github.com/apache/accumulo/blob/e93cb7ac55cac70197fe7df2b7208e1f0a16150d/core/src/main/java/org/apache/accumulo/core/conf/Property.java#L976) binds to the more specific method when building against 1.10.

**Comment by ctubbsii** [comment]
Is it possible to simply drop in commons-configuration 1.10 into the Hadoop class path?

**Comment by keith-turner** [comment]
> Is it possible to simply drop in commons-configuration 1.10 into the Hadoop class path?

I tried replacing hadoops commons config with 1.10 and the job starts, but the map task fail with a weird error.  I am trying to figure out the quickest way to actually get the verify job to run so I can validate the test.

**Comment by ctubbsii** [comment]
I was going to try to cast the parameter to force the binding to the less specific method, without having to downgrade.

**Comment by keith-turner** [comment]
> I was going to try to cast the parameter to force the binding to the less specific method, without having to downgrade.

I cast it to java.util.Map, rebuilt accumulo core, and dropped that into accumulo's lib dir.  The map reduce job started and is currently reading from Accumulo.   With this approach, how will we know if we got everything that matters?

**Comment by mjwall** [comment]
What about going back to commons config 1.6 for the 1.9.3 release? @busbey
any issue with that?  I worry about making it hard for current 1.9.2 users
to upgrade.

**Comment by ctubbsii** [comment]
> > I was going to try to cast the parameter to force the binding to the less specific method, without having to downgrade.
> 
> I cast it to java.util.Map, rebuilt accumulo core, and dropped that into accumulo's lib dir. The map reduce job started and is currently reading from Accumulo. With this approach, how will we know if we got everything that matters?

We might not have... but I think it's low risk. We can always address new stuff as we find it. I'm pretty sure there aren't any, though... the API did not change all that much between 1.x commons-configuration versions. The cast is a simple solution. If it fixes the problem, I suggest we just do that.

**Comment by joshelser** [comment]
What about moving to Hadoop3 and using the new shaded artifacts? Is there desire to keep using Hadoop2?

**Comment by ctubbsii** [comment]
@joshelser This is only for the 1.9 branch, which still supports Hadoop 2 and Hadoop 3. This isn't an issue for the master branch, which has already moved to Hadoop 3.

**Comment by joshelser** [comment]
> This is only for the 1.9 branch, which still supports Hadoop 2 and Hadoop 3

This was the point of me asking such a question. Is it still worth the headache to keep fighting this battle?

**Comment by ctubbsii** [comment]
> > This is only for the 1.9 branch, which still supports Hadoop 2 and Hadoop 3
> 
> This was the point of me asking such a question. Is it still worth the headache to keep fighting this battle?

I think so, yes.

People are, and want to continue, using Accumulo 1.9. And many of those are likely not ready to move to Hadoop 3. We've (implicitly, if not more) committed to supporting Hadoop 2 in 1.9, and I don't think these occasional issues are significant enough to abandon that commitment. Of all the headaches we've had at various points with outdated dependencies in Hadoop and class path conflicts, I think this one is a pretty trivial one, and worth fixing to continue to support Hadoop 2 in 1.9.

**Comment by keith-turner** [comment]
> What about moving to Hadoop3 and using the new shaded artifacts? Is there desire to keep using Hadoop2?

From my experience with Hadoop 3 the shaded artifacts are not used in the map reduce environment and one must still fight the dependency battle.  Is there a way to use the shaded artifacts in Map Reduce?

**Comment by joshelser** [comment]
There is a shaded jar specifically for mapreduce. hadoop-mapreduce-client or similar.

**Comment by mjwall** [comment]
I am unclear where we are heading on this.  Seems like we have 3 options.  Please correct this.

1 - Do a cast like @keith-turner as done, but how we ensure we got them all
2 - Move back to commons config 1.6, haven't heard from @busbey 
3 - Use the Hadoop 3 shaded jar, but that forces a Hadoop upgrade for anyone that wants to go to 1.9.3

Should we start a discussion on dev@a.a.o?

**Comment by ctubbsii** [comment]
@mjwall It sounds like the workaround Keith did will work fine, without a need to downgrade. It's pretty easy to look and see if it's the only time we call that method. I don't think option 3 is viable, since it breaks user expectations for supporting Hadoop 2 so badly, and so last-minute, and violates our previous commitment to make 1.9.x the "bridge" version between Hadoop 2 and 3. But, if somebody wants to argue for that, I think it can/should be further discussed separately from this specific compatibility bug, on the dev list.

**Comment by keith-turner** [comment]
> There is a shaded jar specifically for mapreduce. hadoop-mapreduce-client or similar.

Ok when I was trying to get Accumulo 2.0 to work with map reduce, I went down the path of map reduce classloader dependency isolation options.  I'll look into that and see if I can figure out how to configure map reduce to use it.

**Comment by keith-turner** [comment]
I still don't fully understand what the motivation for moving from ver 1.6 to 1.10 was.  It seems like it would be fine to leave it at 1.6, that is just the minimum version we support and build against.  A user can always bump their runtime version to 1.10 if we build against 1.6.

**Comment by ctubbsii** [comment]
> I still don't fully understand what the motivation for moving from ver 1.6 to 1.10 was. It seems like it would be fine to leave it at 1.6, that is just the minimum version we support and build against. A user can always bump their runtime version to 1.10 if we build against 1.6.

If we only produced source code, binding to the oldest supported API version would work fine... except that:

1. commons-configuration doesn't necessarily have API guarantees, if we bind to the older version, there's no guarantee it will work with the newer versions or vice-versa,
2. the newer versions [fix lots of bugs](https://commons.apache.org/proper/commons-configuration/changes-report.html#a1.10),
3. we're not just releasing the source, we're also including the commons-configuration jar in our "convenience binary" tarball assembly.

Because of 1, we have to do our due diligence to check that it will work with regardless of which we choose. Because of 2, we'd want to upgrade to ensure users aren't set up for failure, because of being forced to use an older, buggy version, because of possible incompatibilities with a newer versions. Because of 3, we have to be careful not to set users up for failure by shipping a known buggy version in our own binary assembly that they're likely to use.

All that said, I do not care if we revert #659 or apply the cast workaround to bind to the correct method. Both solutions seem to work fine. I do have a very slight preference towards keeping the commons-configuration with fewer known bugs, but I'd settle for either solution.

**Comment by ctubbsii** [comment]
Also, FWIW, the incompatibility was introduced with the attempt to get Java 5 generics support in commons-configuration 1.8. Version 1.7 does not have this issue, but still fixes many bugs. We could meet in the middle on that.

Also, @keith-turner , the argument "A user can always bump their runtime version to 1.10 if we build against 1.6." applies to Hadoop also. This issue goes away entirely if a user simply drops-in the updated commons-configuration 1.10 into their Hadoop runtime installation.

**Comment by keith-turner** [comment]
> This issue goes away entirely if a user simply drops-in the updated commons-configuration 1.10 into their Hadoop runtime installation.

I tried that and things blew up.  I did not investigate why :)

**Comment by milleruntime** [comment]
I made a PR for the simplest fix for this: https://github.com/apache/accumulo/pull/1068

**Comment by milleruntime** [comment]
Fixed in #1068
